Refactor and implement debug pretty printing, add tests

DarrenTsung opened the pull request:
Hi @derekdreery, thanks for the useful crate! I wanted to parse the debug output for some functionality in a different crate (https://github.com/colin-kiegel/rust-pretty-assertions/issues/85) and found your reddit post.

My main motivation for forking was to implement debug pretty printing ("{:#?}" - https://doc.rust-lang.org/std/fmt/trait.Debug.html), which I found to be pretty straight-forward by replacing the `fmt::Display`
implementations with `fmt.debug_map()`, `fmt.debug_list()`, etc. It ended up expanding in scope a bit though.

This is a pretty large PR / major refactor, so I understand if you don't have the bandwidth to review it!

Commit message below:

---
ValueKind got renamed to Value and is now the top-level struct. Value
got moved into ValueKind as Value::Struct.

Switch implementation of fmt::Display to fmt::Debug and used the
fmt::Formatter debug_{struct, map, etc} methods, which requires less
custom code and allows for pretty printing the debug output!

Changed Maps to allow for `Value` key, see `hashmap_with_object_keys()`
test for example.

Added `name: Option<String>` for Tuples to fix object tuples.

DarrenTsung committed:
Update code and add tests for matching debug printing functionality

ValueKind got renamed to Value and is now the top-level struct. Value
got moved into ValueKind as Value::Struct.

Switch implementation of fmt::Display to fmt::Debug and used the
fmt::Formatter debug_{struct, map, etc} methods, which requires less
custom code and allows for pretty printing the debug output!

Changed Maps to allow for `Value` key, see `hashmap_with_object_keys()`
test for example.

Added `name: Option<String>` for Tuples to fix object tuples.

derekdreery commented:
Hey @DarrenTsung, thanks for the PR. Yes it might take me a while to get to reviewing it, but I always think it's a good thing to share forks with upstream in case they want to merge the changes. Hopefully you're ok with using your fork in the mean time.

DarrenTsung committed:
Derive Ord for Value

DarrenTsung committed:
Silence unused warnings in tests

DarrenTsung committed:
Convert more handwritten code to nom parsers

DarrenTsung committed:
Add UnquotedRawString variant, remove I128, U128, F64 identification

To support a broader range of use-cases and because we don't need
to actually understand the underlying type of the object to output
the same debug format, we can use this UnquotedRawString variant.

DarrenTsung committed:
Update Cargo.toml + remove evcxr

Giving up on the fork accepting my changes..
